src/lib/ams.c

```diff
@@ -0,0 +1,736 @@
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <sof/lib/ams.h>
+
+LOG_MODULE_REGISTER(ams, CONFIG_SOF_LOG_LEVEL);
+
+DECLARE_SOF_UUID("ams", ams_uuid, 0xea9c4bca, 0x5b7d, 0x48c6,
+		 0x95, 0x86, 0x55, 0x3e, 0x27, 0x23, 0x5b, 0xeb);
+
+DECLARE_TR_CTX(ams_tr, SOF_UUID(ams_uuid), LOG_LEVEL_INFO);
+
+static struct ams_context ctx;
+static struct ams_shared_context shared_ctx;
+
+static struct ams_shared_context __sparse_cache *ams_acquire(struct ams_shared_context *shared)
+{
+	struct coherent __sparse_cache *c = coherent_acquire_thread(&shared->c,
+								    sizeof(*shared));
+
+	return attr_container_of(c, struct ams_shared_context __sparse_cache,
+					c, __sparse_cache);
+}
+
+static void ams_release(struct ams_shared_context __sparse_cache *shared)
+{
+	coherent_release_thread(&shared->c, sizeof(*shared));
+}
+
+static struct uuid_idx __sparse_cache *ams_find_uuid_entry_by_uuid(struct ams_shared_context __sparse_cache *ctx_shared,
+								   uint8_t const *uuid)
+{
+	unsigned int index;
+	struct uuid_idx __sparse_cache *uuid_table = ctx_shared->uuid_table;
+	size_t uuid_table_size = ctx_shared->uuid_table_size;
+
+	if (uuid == NULL)
+		return NULL;
+
+	/* try to find existing entry */
+	for (index = 0; index < uuid_table_size; index++) {
+		if (memcmp(uuid_table[index].message_uuid,
+			   uuid, AMS_MESSAGE_UUID_SIZE) == 0) {
+			return &uuid_table[index];
+		}
+	}
+
+	/* and add new one if needed */
+	for (index = 0; index < uuid_table_size; index++) {
+		if (uuid_table[index].message_type_id == AMS_INVALID_MSG_TYPE) {
+			int ec = memcpy_s(uuid_table[index].message_uuid,
+					  sizeof(uuid_table[index].message_uuid),
+					  uuid, AMS_MESSAGE_UUID_SIZE);
+			if (ec != 0)
+				return NULL;
+
+			uuid_table[index].message_type_id = ++ctx_shared->last_used_msg_id;
+			return &uuid_table[index];
+		}
+	}
+
+	return NULL;
+}
+
+int ams_get_message_type_id(const uint8_t *message_uuid,
+			    uint32_t *message_type_id)
+{
+	struct async_message_service *ams = sof_get()->ams;
+	struct uuid_idx __sparse_cache *uuid_entry;
+	struct ams_shared_context __sparse_cache *shared_c;
+
+	if (ams->ams_context == NULL)
+		return -EINVAL;
+
+	*message_type_id = AMS_INVALID_MSG_TYPE;
+
+	shared_c = ams_acquire(ams->ams_context->shared);
+
+	uuid_entry = ams_find_uuid_entry_by_uuid(shared_c, message_uuid);
+	if (uuid_entry == NULL) {
+		ams_release(shared_c);
+		return -EINVAL;
+	}
+
+	*message_type_id = uuid_entry->message_type_id;
+	ams_release(shared_c);
+
+	return 0;
+}
+
+int ams_find_uuid_index_by_msg_type_id(struct ams_shared_context __sparse_cache *ctx_shared,
+				       uint32_t const message_type_id)
+{
+	struct uuid_idx __sparse_cache *iter;
+	unsigned int idx = 0;
+
+	if (message_type_id == AMS_INVALID_MSG_TYPE)
+		return -EINVAL;
+
+	for (int i = 0; i < ctx_shared->uuid_table_size; i++) {
+		iter = &ctx_shared->uuid_table[i];
+
+		/* no more valid ids */
+		if (iter->message_type_id != AMS_INVALID_MSG_TYPE)
```

Comment:
this means, that this function can only find entries with `iter->message_type_id == AMS_INVALID_MSG_TYPE`. This is probably not what you meant.

Reply:
ack - I think this needs fixed. Btw, might be worth while with some UTs for AMS later on.